### PR TITLE
Revamp guard debug logging

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107532
* #107530
* #107516
* __->__ #107505

The new guard printout looks like this:

```
[DEBUG] GUARDS:
[DEBUG]   ___check_type_id(L['name'], 7605632)                          # if name == "special_attr":  # test/dynamo/test_misc.py:1155 in __getattribute__
[DEBUG]   L['name'] == '_backward_pre_hooks'                            # if name == "special_attr":  # test/dynamo/test_misc.py:1155 in __getattribute__
[DEBUG]   ___check_obj_id(L['self'], 139746432564960)                   # return super().__getattribute__(name)  # test/dynamo/test_misc.py:1157 in __getattribute__
[DEBUG]   ___check_obj_id(L['__class__'], 1451499216)                   # return super().__getattribute__(name)  # test/dynamo/test_misc.py:1157 in __getattribute__
[DEBUG]   ___is_grad_enabled()                                          # _dynamo/output_graph.py:346 in init_ambient_guards
[DEBUG]   not ___are_deterministic_algorithms_enabled()                 # _dynamo/output_graph.py:342 in init_ambient_guards
[DEBUG]   ___is_torch_function_enabled()                                # _dynamo/output_graph.py:350 in init_ambient_guards
[DEBUG]   utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:348 in init_ambient_guards
```

Along with the guards, we also print what line of user code caused the guard to be added, or what line of Dynamo internal code added the guard (if there is no user stack trace, which is typically the case for ambient guards.)

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_dynamo/symbolic_convert.py [hunk 1]
```diff
@@ -2027,46 +2027,49 @@ def __init__(
...
-                    ]
-                    self.output.guards.update(index_guards)
+            self.init_local_index_guards_hack()
 
             self._freevars_ids = dict()
             for name in self.code_options["co_freevars"]:
                 if name in f_locals:
                     self._freevars_ids[name] = id(f_locals[name])
 
+    def init_local_index_guards_hack(self):
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:2037]: Just code motion, so I can get `init_local_index_guards_hack` in the debug print

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -346,6 +332,25 @@ def __init__(
         # presence of torch.no_grad) and there is a graph break.
         self.save_global_state()
 
+    # This gets its own helper function so guards DEBUG logs are more
+    # informative
+    def init_ambient_guards(self):
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:337]: Just code motion, so I can get `init_ambient_guards` in the name

File: torch/utils/_traceback.py [hunk 3]
```diff
@@ -130,21 +130,30 @@ def report_compile_source_on_error():
...
     Shorten a source filepath, under the assumption that anything under torch/
     directory is "obvious" and doesn't need to be shown to user.
     """
+    if base is None:
+        base = os.path.dirname(os.path.dirname(__file__))
     # Truncate torch/foo.py to foo.py
-    prefix = os.path.commonprefix([fn, os.path.join(os.path.dirname(os.path.dirname(__file__)), "")])
-    return fn[len(prefix):]
+    try:
+        prefix = os.path.commonpath([fn, base])
```
Comment (ezyang) [inline torch/utils/_traceback.py:142]: This is fixing a bug where 'test/test_foo.py' got shortened to 'est/test_foo.py'

File: torch/_guards.py [hunk 4]
```diff
@@ -148,10 +148,6 @@ class Guard:
     stack = None
     user_stack = None
 
-    def __post_init__(self):
-        self.stack = CapturedTraceback.extract(skip=2)
-        self.user_stack = TracingContext.extract_stack()
```
Comment (ezyang) [inline torch/_guards.py:153]: @jansel told me and I agree that the construction site of guards is not as useful as when they are installed.

File: torch/_dynamo/guards.py [hunk 5]
```diff
@@ -1007,7 +1007,7 @@ def convert(size_or_stride):
         guard_body, pycode = build_guard_function(unique_code_parts, make_guard_fn_args)
 
         if os.environ.get("TORCHDYNAMO_PRINT_GUARDS", None) == "1":
-            print("GUARDS", guard_body)
+            print("GUARDS\n", guard_body)
```
Comment (ezyang) [inline torch/_dynamo/guards.py:1010]: I am going to get to this next, but this PR already is pretty big.

**Review by mlazos — APPROVED** [review]

File: torch/_guards.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_guards.py:153]: Why not both? Both have value imo.

The way I wanted to do guard creation logging is to add a `reason` field - most initial guard association/creation sites are manual/have some reason that we hold in tribal knowledge.

File: torch/utils/_traceback.py (hunk 3 above)
Comment (voznesenskym) [reply torch/utils/_traceback.py:142]: lol

**Review by voznesenskym — APPROVED** [review]

**Review by anijain2305 — APPROVED** [review]

File: torch/_guards.py (hunk 4 above)
Comment (ezyang) [reply torch/_guards.py:153]: It's a signal-to-noise problem. Yes, I *could* report the allocation site and the installation site but you're doubling the amount of information you have to sift through. A lot of the trickiness here is making a good decision about what to give to the user by default. That being said, I'm amenable to changing up what we record, it's not a hard change, we can tweak it as we get experience.

**Pull request closed without merging**